src/shims/fs.rs

```diff
@@ -257,6 +264,196 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.try_unwrap_io_result(result)
     }
 
+    fn statx(
+        &mut self,
+        dirfd_op: OpTy<'tcx, Tag>,
+        pathname_op: OpTy<'tcx, Tag>,
+        flags_op: OpTy<'tcx, Tag>,
+        _mask_op: OpTy<'tcx, Tag>,
+        statxbuf_op: OpTy<'tcx, Tag>
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("statx")?;
+
+        let statxbuf_scalar = this.read_scalar(statxbuf_op)?.not_undef()?;
+        let pathname_scalar = this.read_scalar(pathname_op)?.not_undef()?;
+
+        // If the statxbuf or pathname pointers are null, the function fails with `EFAULT`.
+        if this.is_null(statxbuf_scalar)? || this.is_null(pathname_scalar)? {
+            let efault = this.eval_libc("EFAULT")?;
+            this.set_last_error(efault)?;
+            return Ok(-1);
+        }
+
+        // If statxbuf is not null we need to produce a proper `MemPlace` to write the results of
+        // the function to it. This is needed because all the `statx` parameters are provided with
+        // an `i64` layout instead of the proper type layout.
+        let statxbuf_ptr = this.force_ptr(statxbuf_scalar)?;
+        // This long path is required instead of `libc::statx` because the `statx` struct and
+        // function share the same namespace.
+        let statx_ty = this
+            .resolve_path(&["libc", "unix", "linux_like", "linux", "gnu", "statx"])?
+            .ty(*this.tcx);
+        let statxbuf_ty = this.tcx.mk_mut_ptr(statx_ty);
+        let statxbuf_layout = this.layout_of(statxbuf_ty)?;
+        let statxbuf_imm = ImmTy::from_scalar(Scalar::Ptr(statxbuf_ptr), statxbuf_layout);
```

Comment:
So you went through the trouble of calling `force_ptr` to get a pointer just to turn it back into a `Scalar` here? That is odd.

Reply:
I couldn't find another way of obtaining an `ImmTy` from `statxbuf_op`. How do you think it should be done?